### PR TITLE
FEATURE: implementing solver classes for python lib

### PR DESCRIPTION
**Pull request by chemardes**
**The following PR addressing Issues #3, #4 , #7** 

Idk if this is the best way to do it , but this was my interpretation of using classes for the python library, comments are welcome!

@alphaville, @debdal6 

Changes:
- Utilised classes to format the code
- Made use of the Strategy Design pattern (https://refactoring.guru/design-patterns/strategy/python/example) to dynamically change the methods used to solve the equation
- Changed the way I calculated dx and dt (used Debdal's approach as it's much cleaner)
- Fixed initial and boundary conditions

Things to do: 
- Need to add comments
- Need to implement error handling
- Need to add tests (checking the error between methods + unittests)
- Maybe make a utils file?

### CODE REVIEW

File: pdesolvers/heat_equation_1d/solver/explicit_solver.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+
+class HeatEquationExplicitSolver(SolverStrategy):
+
+    def solve(self, context):
+
+        x = context._generate_x_grid()
+        dx = x[1] - x[0]
+
+        dt_max = 0.5 * (dx**2) / context._get_k()
```
Comment (chemardes) [inline pdesolvers/heat_equation_1d/solver/explicit_solver.py:13]: would it be better if i did it like this (e.g. automatically calculating the dt within the method to ensure stability)

or

do a check and tell the user that the inputted values are invalid (e.g. too large)?

File: pdesolvers/heat_equation_1d/heat_equation.py [hunk 2]
```diff
@@ -0,0 +1,129 @@
+import numpy as np
+from matplotlib import pyplot as plt
+
+from pdesolvers.heat_equation_1d.strategy.solver_strategy import SolverStrategy
+
+
+class HeatEquation:
```
Comment (alphaville) [inline pdesolvers/heat_equation_1d/heat_equation.py:7]: It seems that this class is a `HeatEquationSolver` rather than a `HeatEquation`. Is that right?

File: pdesolvers/heat_equation_1d/heat_equation.py [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+
+    def __init__(self, length, x_nodes, time, t_nodes, k, solver: SolverStrategy = None):
```
Comment (alphaville) [inline pdesolvers/heat_equation_1d/heat_equation.py:9]: Typically you would have a `Solver` class and you provide a problem to it, rather than a problem class to which you provide a solver. This is how other solvers work (see for example https://alphaville.github.io/optimization-engine/, https://www.cvxpy.org/, etc)

File: pdesolvers/heat_equation_1d/heat_equation.py [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+        if self.__right_boundary_temp is not None:
+            err = np.abs(self.__right_boundary_temp(0) - self.__initial_temp(0))
+            assert err < 1e-12
+
+    @staticmethod
+    def __validate_callable(func):
+        if not callable(func):
+            raise ValueError("Temperature conditions must be a callable function")
+
+    def _generate_x_grid(self):
```
Comment (alphaville) [inline pdesolvers/heat_equation_1d/heat_equation.py:90]: Why do these methods start with a single underscore?

File: pdesolvers/heat_equation_1d/main.py [hunk 5]
```diff
@@ -0,0 +1,22 @@
+import numpy as np
+from heat_equation import HeatEquation
+from pdesolvers.heat_equation_1d.solver.cn_solver import HeatEquationCNSolver
+from pdesolvers.heat_equation_1d.solver.explicit_solver import HeatEquationExplicitSolver
+
+
+def main():
+
+    equation = (((HeatEquation(1, 100,30,10000, 0.01)
```
Comment (alphaville) [inline pdesolvers/heat_equation_1d/main.py:9]: I'm a little confused with the parentheses here.

File: pdesolvers/heat_equation_1d/heat_equation.py (hunk 4 above)
Comment (alphaville) [reply pdesolvers/heat_equation_1d/heat_equation.py:90]: I guess you tried to make "friend" method (as in C++), which are not supported in Python. A different design (Problem --> Solver) would simplify things.

**Review by alphaville — COMMENTED** [review]
Very good. I just have some doubts about this design; see my comments below.

File: pdesolvers/heat_equation_1d/main.py (hunk 5 above)
Comment (chemardes) [reply pdesolvers/heat_equation_1d/main.py:9]: it just allows me to format the .function_call

File: pdesolvers/heat_equation_1d/heat_equation.py (hunk 4 above)
Comment (chemardes) [reply pdesolvers/heat_equation_1d/heat_equation.py:90]: i was trying to access the methods in HeatEquation in the solver classes, but then realised the user would be able to access those methods too, so i added the underscore to make it protected

File: pdesolvers/heat_equation_1d/solver/cn_solver.py [hunk 6]
```diff
@@ -65,4 +66,26 @@ def __build_tridiagonal_matrix(a, b, c, nodes):
 
         matrix = csc_matrix(matrix)
 
-        return matrix
\ No newline at end of file
+        return matrix
+
+    def plot(self):
```
Comment (chemardes) [inline pdesolvers/heat_equation_1d/solver/cn_solver.py:71]: i feel like this plot function could be moved elsewhere as it is a shared function (any suggestions?)

File: pdesolvers/heat_equation_1d/solver/explicit_solver.py [hunk 7]
```diff
@@ -1,32 +1,59 @@
...
 
 
 class HeatEquationExplicitSolver(SolverStrategy):
 
-    def solve(self, context):
 
-        x = context._generate_x_grid()
+    def solve(self, equation : HeatEquation):
+
+        self.equation = equation
```
Comment (chemardes) [inline pdesolvers/heat_equation_1d/solver/explicit_solver.py:13]: don't think this is the best way to go about with this - i did this so that i could access the equation in the plot() function

i was initially gonna do an init so that both solve and plot has access to the equation, but then it would be like:

solver = Solver(HeatEquationCNSolver(equation))
solver.solve().plot()

and not 

solver = Solver(HeatEquationCNSolver())
solver.solve(equation).plot()

maybe im overcomplicating things

File: pdesolvers/heat_equation_1d/solver/cn_solver.py (hunk 6 above)
Comment (alphaville) [reply pdesolvers/heat_equation_1d/solver/cn_solver.py:71]: You plot the solution, not the solver, so it can be moved to a `Solution` class. Then every solver can return an instance of `Solution` that can be plotted.
